### PR TITLE
feat(button): submit from outside of form

### PR DESCRIPTION
**Pull request by postnerd**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?
Buttons with type "submit" will only work from inside forms.

Issue URL: https://github.com/ionic-team/ionic-framework/issues/21194

## What is the new behavior?
Now you can add a "form" prop to every <ion-button> that will be used, if the button is placed outside of a form. This can be helpful, if the button is added to a header/footer toolbar but should be responsible for a specific form.

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Not sure, if the docs have to be updated. If so, I'm happy to do so. Maybe you could guide me to the right place.

### CODE REVIEW

File: core/src/components/button/button.tsx
Comment (sean-perkins) [inline core/src/components/button/button.tsx:127]:
```suggestion
   * The HTML form element or form element id. Used to submit a form when the button is not a child of the form.
```

**Review by sean-perkins — CHANGES_REQUESTED** [review]
Small suggestion on the documentation (will require a new `npm run build` for the generated outputs). 

We will also need Playwright tests for this feature.

For example, in `core/src/components/button/test/form-reference` create a new `button.e2e.ts` file.

You will need to test 3 conditions:
1. Form with an ID
2. Form as a reference
3. No form (querying the closest form). 

For these, setting the HTML mark-up per test is probably easiest. 

I did not run these tests locally, but something like this:
```ts
test.describe('button: form', () => {

  test('should submit the form by id', async({ page }) => {
    await page.setContent(`
      <form id="myForm"></form>
      <ion-button form="myForm" type="submit">Submit</ion-button>
    `);

    const submitEvent = await page.spyOnEvent('submit');
    
    await page.click('ion-button');

    expect(submitEvent).toHaveReceivedEvent();
  });

  test('should submit the form by reference', async({ page }) => {
    await page.setContent(`
      <form></form>
      <ion-button type="submit">Submit</ion-button>
      <script>
        const form = document.querySelector('form');
        const button = document.querySelector('ion-button');

        button.form = form;
      </script>
    `);

    const submitEvent = await page.spyOnEvent('submit');
    
    await page.click('ion-button');

    expect(submitEvent).toHaveReceivedEvent();
  });

  test('should submit the closest form', async({ page }) => {
    await page.setContent(`
    <form>
      <ion-button type="submit">Submit</ion-button>
    </form>
  `);

  const submitEvent = await page.spyOnEvent('submit');
  
  await page.click('ion-button');

  expect(submitEvent).toHaveReceivedEvent();
  });
});
```

Let me know if you have questions.

**Comment by postnerd** [comment]
@sean-perkins Thanks for the support. But I can't get the tests to run.

I've tried all possible commands for testing `npm run test.screenshot` or `npm run test` or `npm run run test.e2e` but only the plain _e2e.ts_ files will be executed not the _button.e2e.ts_ files.

`npm test src/components/button/test/basic/button.e2e.ts ` will run into the following error message `Pattern: src/components/button/test/basic/button.e2e.ts - 0 matches`.

**Comment by liamdebeasi** [comment]
@postnerd We are in the middle of a transition between two E2E testing systems. The `npm run test.e2e` command is for the old system. You should do `npx playwright test src/components/button/test/basic` to run the new Playwright tests.

**Review by postnerd — COMMENTED** [review]
Added the required tests.

Please let me know if there is anything else to do.

(Messed up this one commit when adding the tests with a wrong commit message.) Since this is such a small PR I could just create another branch and PR with just one commit if you want me to.)

**Comment by sean-perkins** [comment]
The commit message is ok here. We will end up squashing the commits anyways, so only the topmost description (the PR title) will matter. 

I approved CI to run to verify the tests. I don't see any other items requiring changes on my end. 

If CI passes, I'll add the team and we can discuss the release target for this feature.

**Comment by postnerd** [comment]
@sean-perkins Is there anything  to do from my side because of the failed tests?

**Comment by sean-perkins** [comment]
@postnerd merge the latest change from `main` into this branch, run `npm run build` and commit any changes (if any).

**Comment by postnerd** [comment]
Merged again (after @liamdebeasi already merged) because the branch was again one commit behind. 

`npm run build` didn't bring anything up related to this pull request.

Only one change to _angular/src/directives/proxies-list.ts_ that I've ignored before, because it looks like an unrelated style fix (a strange one, adding a line at the beginning of the file ...).

Screenshot 2022-09-19 at 21 49 31
Screenshot 2022-09-19 at 21 49 39

**Comment by sean-perkins** [comment]
Yeah, you can ignore that diff. Running `npm run prettier` from the `/angular` folder will overwrite that change and reformat the file with the current state. 

I re-triggered CI to run.

**Comment by sean-perkins** [comment]
@postnerd update on our end: The team has drafted and approved a design document around this feature. The only item that is outstanding as a result, is we want to introduce a developer warning when the form cannot be found, when using the `form` syntax. 

I can snag this scope & can commit it directly to your fork. 

We will target this feature for 6.3 after I make that change & then the team can do a final review.

**Review by sean-perkins — APPROVED** [review]

**Review by amandaejohnston — APPROVED** [review]
LGTM once the comment is addressed 👍

File: core/src/components/button/button.tsx
Comment (liamdebeasi) [inline core/src/components/button/button.tsx:210]: This warning is a bit confusing because it implies that the form element is not in the DOM at all. However, it could be the case that the element is but is not an HTMLFormElement.

File: core/src/components/button/button.tsx
Comment (liamdebeasi) [inline core/src/components/button/button.tsx:222]: Why was this logic placed outside `findForm`? I would have expected it to be included in that function as the fallback.

File: core/src/components/button/test/form-reference/button.e2e.ts
Comment (liamdebeasi) [inline core/src/components/button/test/form-reference/button.e2e.ts:64]: This check seems brittle. The moment we change the contents of the warning, this test will fail. Are you trying to check that a warning is logged or are you trying to check what the contents of the warning are?

File: core/src/components/button/test/form-reference/index.html
Comment (liamdebeasi) [inline core/src/components/button/test/form-reference/index.html:25]:
```suggestion
      <ion-content class="ion-padding ion-text-center" id="content">
```

**Review by liamdebeasi — CHANGES_REQUESTED** [review]

File: core/src/components/button/button.tsx
Comment (sean-perkins) [reply core/src/components/button/button.tsx:222]: This was done for this check here: https://github.com/ionic-team/ionic-framework/pull/25913/files#diff-8b0e185d2b3a4ceee4ec690a5f8ff2992ef3720594562f34bf2cae17f05254ffR198

If the user provides a `form` element, but the `findForm()` function returned the fallback of querying the closest form, there could be instances in applications where the button is attaching to the wrong form element, that the user is not expecting. 

If the `findForm()` function only returns the result of querying the provided `form` or returning the element directly (if `form` is an `HTMLFormElement`), then we can check if that is not set and warn appropriately.

File: core/src/components/button/test/form-reference/button.e2e.ts
Comment (sean-perkins) [reply core/src/components/button/test/form-reference/button.e2e.ts:64]: Both. Although I could just check for `[Ionic Warning]` if we are concerned with the message contents being updated and needing to update the test over time.

Comment (liamdebeasi) [reply core/src/components/button/test/form-reference/button.e2e.ts:64]: I'm fine keeping this as is for now. Long term it may be good to have a warning/error ID like "MISSING_FORM_ELEMENT" that we could check, but that needs design work first and should not block this PR.

File: core/src/components/button/button.tsx
Comment (sean-perkins) [reply core/src/components/button/button.tsx]: Updated the warning message. Let me know if it is more clear or requires additional adjustment.

**Review by liamdebeasi — APPROVED** [review]

**Comment by sean-perkins** [comment]
Great work here @postnerd 👍 thanks for the contribution!

**Pull request merged**
